[Merged by Bors] - feat(topology): tiny new lemmas

PatrickMassot opened the pull request:
from LTE



---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

PatrickMassot committed:
feat(topology): tiny new lemmas

from LTE

eric-wieser commented on src/topology/constructions.lean:
```diff
@@ -64,6 +64,15 @@ instance Pi.topological_space {β : α → Type v} [t₂ : Πa, topological_spac
+  {a : α} (hab : quotient.mk a = q) (hs : V ∈ 𝓝 q) : quotient.mk ⁻¹' V ∈ 𝓝 a :=
+preimage_mem_nhds_of_coinduced rfl hab hs
+
 /-- The image of a dense set under `quotient.mk` is a dense set. -/
 lemma dense.quotient [setoid α] [topological_space α] {s : set α} (H : dense s) :
   dense (quotient.mk '' s) :=
```
Based on the name of this lemma, perhaps `quotient.is_open_preimage` should be called `is_open.quotient_preimage` to match, and so that dot notation works on `h : is_open V` - what are your thoughts on that?

PatrickMassot replied:
Ok, why not.

PatrickMassot committed:
Rename lemma following Eric's suggestion

eric-wieser commented on src/topology/order.lean:
```diff
@@ -294,6 +294,21 @@ lemma is_open_coinduced {t : topological_space α} {s : set β} {f : α → β}
+
+lemma preimage_mem_nhds_of_coinduced [topological_space α] [topological_space β] {π : α → β}
+  (h : ‹topological_space β› = topological_space.coinduced π ‹_›) {s : set β} {b : β}
+  {a : α} (hab : π a = b) (hs : s ∈ 𝓝 b) : π ⁻¹' s ∈ 𝓝 a :=
```
Does it make sense to eliminate `b` here?
```suggestion
lemma preimage_mem_nhds_of_coinduced [topological_space α] [topological_space β] {π : α → β}
  (h : ‹topological_space β› = topological_space.coinduced π ‹_›) {s : set β}
  {a : α} (hs : s ∈ 𝓝 (π a)) : π ⁻¹' s ∈ 𝓝 a :=
```

eric-wieser commented on src/topology/constructions.lean:
```diff
@@ -64,6 +64,15 @@ instance Pi.topological_space {β : α → Type v} [t₂ : Πa, topological_spac
+
+lemma quotient.preimage_mem_nhds [topological_space α] [s : setoid α]
+  {V : set $ quotient s} {q : quotient s}
+  {a : α} (hab : quotient.mk a = q) (hs : V ∈ 𝓝 q) : quotient.mk ⁻¹' V ∈ 𝓝 a :=
```
On the same theme as the one below, this lemma could eliminate `q` as
```suggestion
  {a : α} (hs : V ∈ 𝓝 (quotient.mk a)) : quotient.mk ⁻¹' V ∈ 𝓝 a :=
```

PatrickMassot replied:
This will mean an extra rewrite when using the lemma, but it's true that it makes the statement cleaner, so let's do that.

eric-wieser commented on src/topology/order.lean:
```diff
@@ -294,6 +294,21 @@ lemma is_open_coinduced {t : topological_space α} {s : set β} {f : α → β}
   @is_open β (topological_space.coinduced f t) s ↔ is_open (f ⁻¹' s) :=
 iff.rfl
 
```
This lemma matches the style of `is_open_coinduced` above, so is probably a good idea:
```suggestion

lemma preimage_mem_nhds_coinduced [topological_space α] {π : α → β}
  {s : set β} {a : α} :
  s ∈ @nhds _ (topological_space.coinduced π ‹_›) (π a) → π ⁻¹' s ∈ 𝓝 a :=
begin
  letI := topological_space.coinduced π ‹_›,
  rw [mem_nhds_sets_iff, mem_nhds_sets_iff],
  rintros ⟨V, hVs, V_op, mem_V⟩,
  exact ⟨π ⁻¹' V, set.preimage_mono hVs, is_open_coinduced.mp V_op, mem_V⟩
end

```

PatrickMassot replied:
No, that's definitely a bad idea. It makes the lemma less general. Your version will work only when the topology on `β` is definitionally equal to the topology pushed by `π` from `α`.

eric-wieser replied:
Sure, but if it's not definitionally equal they can extract a propositional goal with `convert preimage_mem_nhds_coinduced` or similar.

Note that you [can prove the "general"](https://github.com/leanprover-community/mathlib/pull/7554/files#r628885015) version with this lemma + `subst`. So I'd argue that saying this version is less general is somewhat like saying `add_comm : a + b = b + a` is less general than `add_comm (hc : c = a + b) (hd : d = b + a) : c = d`.

PatrickMassot replied:
I'm not convinced at all. With that reasoning you can "prove" anything using `convert iff.rfl` or whatever.

PatrickMassot replied:
Let's take a specific example from `data.nat.basic`. Do you suggest we replace
```lean
theorem eq_one_of_mul_eq_one_right {m n : ℕ} (H : m * n = 1) : m = 1 :=
eq_one_of_dvd_one ⟨n, H.symm⟩
```
by
```lean
theorem eq_one_of_mul_eq_one_right_eric {m : ℕ} (H : m = 1) : m = 1 := H

theorem eq_one_of_mul_eq_one_right {m n : ℕ} (H : m * n = 1) : m = 1 :=
begin
  convert eq_one_of_mul_eq_one_right_eric H,
  rw mul_comm at H,
  rw [eq_one_of_dvd_one ⟨m, H.symm⟩, mul_one]
end
```

eric-wieser replied:
That's not an accurate characterization of my argument. My argument is that if you have a lemma of the form:
```lean
lemma foo {x y z} {p f} (h : x = f y z) : p x y z :=
```
then you should instead have stated it as follows,
```lean
lemma foo' {y z} {p f} : p (f y z) y z :=
```
or at least, proved this version first and provided the messier `foo` later as `subst h; exact foo` only if you have a convincing use-case for it.

That doesn't apply to the `eq_one_of_mul_eq_one_right` snippet you posted, because `H : m * n = 1` is not in the form `y = f x` for some variable `y`.

eric-wieser commented on src/topology/order.lean:
```diff
@@ -294,6 +294,21 @@ lemma is_open_coinduced {t : topological_space α} {s : set β} {f : α → β}
+  rw mem_nhds_sets_iff,
+  exact ⟨π ⁻¹' V, set.preimage_mono hVs, is_open_preimage_of_coinduced h V_op,
+         by rwa ← hab at mem_V⟩
+end
```
With the above suggestion, this becomes:
```suggestion
begin
  subst hab,
  substI h,
  exact preimage_mem_nhds_coinduced hs,
end
```

PatrickMassot committed:
Apply suggestions from code review

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

eric-wieser commented on src/topology/order.lean:
```diff
@@ -294,6 +294,21 @@ lemma is_open_coinduced {t : topological_space α} {s : set β} {f : α → β}
+  rcases mem_nhds_sets_iff.mp hs with ⟨V, hVs, V_op, mem_V⟩,
+  rw mem_nhds_sets_iff,
+  exact ⟨π ⁻¹' V, set.preimage_mono hVs, is_open_preimage_of_coinduced h V_op,
+         by rwa ← hab at mem_V⟩
```
This probably fixes the build:
```suggestion
  exact ⟨π ⁻¹' V, set.preimage_mono hVs, is_open_preimage_of_coinduced h V_op, mem_V⟩
```

PatrickMassot replied:
Sorry, I didn't see this message before fixing it.

PatrickMassot committed:
fix build

eric-wieser commented on src/topology/constructions.lean:
```diff
@@ -64,6 +64,15 @@ instance Pi.topological_space {β : α → Type v} [t₂ : Πa, topological_spac
 
+lemma is_open.quotient_preimage [topological_space α] [s : setoid α]
+  {V : set $ quotient s} (h : is_open V) : is_open (quotient.mk ⁻¹' V) :=
+is_open_preimage_of_coinduced rfl h
```
```suggestion
lemma is_open.quotient_preimage [topological_space α] [s : setoid α]
  {V : set $ quotient s} (h : is_open V) : is_open (quotient.mk ⁻¹' V) :=
is_open_coinduced.mp h
```
which suggests you might want to also provide this as an `iff`. Then again, `is_open_coinduced` is already syntactically equal to the `iff` you'd end up writing.

PatrickMassot committed:
remove lemmas

the only goal is to reduce the number of bikeshedding targets while
keeping the crucial lemma

jcommelin reviewed (approved):
This stuff looks fine to me. Let's merge this now. The other stuff about which there was some discussion can be done in another PR if needed.

bors merge

eric-wieser commented on src/topology/constructions.lean:
```diff
@@ -64,6 +64,11 @@ instance Pi.topological_space {β : α → Type v} [t₂ : Πa, topological_spac
 
+lemma quotient.preimage_mem_nhds [topological_space α] [s : setoid α]
+  {V : set $ quotient s} {q : quotient s}
+  {a : α} (hs : V ∈ 𝓝 (quotient.mk a)) : quotient.mk ⁻¹' V ∈ 𝓝 a :=
```
to fix the linter
```suggestion
  {V : set $ quotient s} {a : α} (hs : V ∈ 𝓝 (quotient.mk a)) : quotient.mk ⁻¹' V ∈ 𝓝 a :=
```

eric-wieser committed:
Update src/topology/constructions.lean
